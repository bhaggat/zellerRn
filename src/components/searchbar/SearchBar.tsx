import React, {useState, useCallback, Dispatch, SetStateAction} from 'react';
import {View, TextInput, StyleSheet, ActivityIndicator} from 'react-native';
import {debounce} from '../../services/utils';

interface SearchBoxProps {
  onSearch: Dispatch<SetStateAction<string>>;
  isLoading?: boolean;
}

export default function SearchBar({onSearch, isLoading}: SearchBoxProps) {
  const [search, setSearch] = useState('');

  const throttledSearch = useCallback(
    debounce((search: string) => {
      onSearch(search);
    }),
    [],
  );

  const handleChangeText = (text: string) => {
    setSearch(text);
    throttledSearch(text);
  };

  return (
    <View style={styles.container}>
      <TextInput
        style={styles.input}
        placeholderTextColor={'black'}
        placeholder="Search..."
        value={search}
        onChangeText={handleChangeText}
        clearButtonMode="always"
      />
      {isLoading && search && (
        <ActivityIndicator
          size="small"
          color="#0000ff"
          testID="activity-indicator"
        />
      )}
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    alignItems: 'center',
    borderWidth: 1,
    borderRadius: 8,
    paddingHorizontal: 8,
    paddingVertical: 4,
    backgroundColor: '#fff',
    borderColor: '#ccc',
    width: '100%',
  },
  input: {
    flex: 1,
    height: 40,
    fontSize: 16,
  },
});
