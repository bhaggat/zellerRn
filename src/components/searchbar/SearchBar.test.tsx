import React from 'react';
import {render, fireEvent, act} from '@testing-library/react-native';
import SearchBar from './SearchBar';

jest.mock('../../services/utils', () => ({
  debounce: jest.fn(fn => fn),
}));

describe('SearchBar Component', () => {
  const onSearchMock = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  describe('Snapshot', () => {
    test('matches', () => {
      const tree = render(<SearchBar onSearch={onSearchMock} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
  describe('Render', () => {
    test('renders correctly with initial state', () => {
      const {getByPlaceholderText, queryByTestId} = render(
        <SearchBar onSearch={onSearchMock} />,
      );

      expect(getByPlaceholderText('Search...')).toBeTruthy();
      expect(queryByTestId('activity-indicator')).toBeNull();
    });
  });
  describe('Interactions', () => {
    test('updates the search input value on typing', () => {
      const {getByPlaceholderText} = render(
        <SearchBar onSearch={onSearchMock} />,
      );

      const input = getByPlaceholderText('Search...');
      fireEvent.changeText(input, 'test query');

      expect(input.props.value).toBe('test query');
    });

    test('calls onSearch with the correct value after debounce', () => {
      const {getByPlaceholderText} = render(
        <SearchBar onSearch={onSearchMock} />,
      );

      const input = getByPlaceholderText('Search...');
      fireEvent.changeText(input, 'debounced query');

      expect(onSearchMock).toHaveBeenCalledWith('debounced query');
    });

    test('shows the activity indicator when isLoading is true', () => {
      const {getByTestId, getByPlaceholderText, queryByTestId} = render(
        <SearchBar onSearch={onSearchMock} isLoading={true} />,
      );

      const input = getByPlaceholderText('Search...');
      fireEvent.changeText(input, 'debounced query');

      expect(getByTestId('activity-indicator')).toBeTruthy();
      expect(queryByTestId('activity-indicator')).not.toBeNull();
    });

    test('hides the activity indicator when isLoading is false', () => {
      const {queryByTestId} = render(
        <SearchBar onSearch={onSearchMock} isLoading={false} />,
      );

      expect(queryByTestId('activity-indicator')).toBeNull();
    });

    test('does not show activity indicator when search input is empty', () => {
      const {queryByTestId} = render(
        <SearchBar onSearch={onSearchMock} isLoading={true} />,
      );

      expect(queryByTestId('activity-indicator')).toBeNull();
    });
  });
});
