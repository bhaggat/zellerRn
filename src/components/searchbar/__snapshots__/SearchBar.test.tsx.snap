// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchBar Component Snapshot Matched 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "borderColor": "#ccc",
      "borderRadius": 8,
      "borderWidth": 1,
      "flexDirection": "row",
      "paddingHorizontal": 8,
      "paddingVertical": 4,
      "width": "100%",
    }
  }
>
  <TextInput
    clearButtonMode="always"
    onChangeText={[Function]}
    placeholder="Search..."
    placeholderTextColor="black"
    style={
      {
        "flex": 1,
        "fontSize": 16,
        "height": 40,
      }
    }
    value=""
  />
</View>
`;

exports[`SearchBar Component Snapshot matches 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "borderColor": "#ccc",
      "borderRadius": 8,
      "borderWidth": 1,
      "flexDirection": "row",
      "paddingHorizontal": 8,
      "paddingVertical": 4,
      "width": "100%",
    }
  }
>
  <TextInput
    clearButtonMode="always"
    onChangeText={[Function]}
    placeholder="Search..."
    placeholderTextColor="black"
    style={
      {
        "flex": 1,
        "fontSize": 16,
        "height": 40,
      }
    }
    value=""
  />
</View>
`;
